src/test/java/org/sonatype/nexus/ci/nxrm/ComponentUploaderNxrm3Test.groovy

```diff
@@ -150,6 +152,61 @@ class ComponentUploaderNxrm3Test
           [new MavenAsset(null, 'classifier', 'extension'), new MavenAsset(null, 'env-classifier', 'env-extension')]
   }
 
+  @WithoutJenkins
+  @Unroll
+  def 'upload component groupId: #groupId, artifactId: #artifactId, version: #version, tag: #tagName'() {
+    setup:
+      def client = Mock(RepositoryManagerV3Client)
+      def nxrmConfiguration = new Nxrm3Configuration('id', 'internalId', 'displayName', 'foo', 'credId')
+      run.getEnvironment(_ as TaskListener) >> new EnvVars([:])
+      client.getTag(_ as String) >> Optional.empty()
+
+      ComponentUploaderNxrm3 mockComponentUploader =
+          Spy([constructorArgs: [nxrmConfiguration, run, taskListener]] as Map<String, Object>,
+              ComponentUploaderNxrm3) {
+            it.getRepositoryManagerClient(nxrmConfiguration) >> client
+          }
+      def publisher = Mock(NexusPublisher)
+      def tempFile = File.createTempFile("temp", ".tmp")
+      tempFile.deleteOnExit()
+
+      def coordinate = new MavenCoordinate(groupId, artifactId, version, packaging)
+      def filePath = new FilePath(tempFile.getParentFile())
+      publisher.packages >> [
+          new MavenPackage(coordinate, [new MavenAsset(tempFile.name, classifier, extension)])
+      ]
+      String repo = null
+      Component component = null
+      String tag = null
+
+    when:
+      mockComponentUploader.uploadComponents(publisher, filePath, tagName)
+      tempFile.delete()
+
+    then:
+      1 * client.upload(*_) >> { args ->
+        repo = args[0]
+        component = args[1]
+        tag = args[2]
+      }
+
```

Comment:
Probably should verify the createTag is called as well given the setup:
`if (tagName) {
        1 * client.createTag(tagName)
      }`

not sure if there's a better way to do conditional checks with groovy/spock